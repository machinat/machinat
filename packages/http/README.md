# Http Module

This module provide HTTP listening capability to all the services require it.
For example, to serve a webhook for subscribing events from chat platform.

## Install

```bash
npm install @sociably/core @sociably/http
# or with yarn
yarn add @sociably/core @sociably/http
```

## Docs

Check the [package reference](https://sociably.js.org/api/modules/http.html).

## Setup

Use `listenOptions` to set the network options for [`server.listen()`](https://nodejs.org/dist/latest/docs/api/net.html#net_server_listen_options_callback).

```js
import Sociably from '@sociably/core';
import Http from '@sociably/http';

const app = Sociably.createApp({
  modules: [
    Http.initModule({
      listenOptions: {
        port: 8080,
        host: 'localhost',
      }
    }),
  ],
});
```

## Usage

Provide `Http.RequestRouteList` or `Http.UpgradeRouteList` (for [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API))
to register RPC style routes. All the requests under the registered `path` will
be received by the handler. You can also use `default: true` to catch all
unmatch requests.

```js
const app = Sociably.createApp({
  modules: [
    Http.initModule({ /* ... */ }),
  ],
  service: [
    { // route listening to /api
      provide: Http.RequestRouteList,
      withValue: {
        name: 'my_api',
        path: '/api',
        handler: (req, res) => {
          // handle http requests
        },
      }
    },
    { // default request route
      provide: Http.RequestRouteList,
      withValue: {
        name: 'default',
        default: true,
        handler: (req, res) => {
          // catch requests not matching any route
        },
      }
    },
    {  // handle WebSocket connection
      provide: Http.UpgradeRouteList,
      withValue: {
        name: 'web_socket',
        path: '/',
        handler: (req, head, socket) => {
          // handle http upgrade requests
        },
      }
    },
  ],
});
```
