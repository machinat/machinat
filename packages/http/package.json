{
  "name": "@machinat/http",
  "version": "0.5.0",
  "author": "Patrick Liu <zeldalink0515@gmail.com>  (https://github.com/lrills)",
  "repository": {
    "type": "git",
    "url": "https://github.com/machinat/machinat.git",
    "directory": "packages/http"
  },
  "homepage": "https://github.com/machinat/machinat#readme",
  "bugs": "https://github.com/machinat/machinat/issues",
  "main": "lib/index.js",
  "exports": {
    ".": "./lib/index.js",
    "./webhook": "./lib/webhook/index.js"
  },
  "files": [
    "lib/*",
    "README.md",
    "polyfill-exports.js"
  ],
  "scripts": {
    "postinstall": "node ./polyfill-exports.js",
    "prepack": "make -f ../../package.mk prepack"
  },
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "invariant": "^2.2.4",
    "raw-body": "^2.4.1",
    "thenifiedly": "^0.2.1"
  },
  "peerDependencies": {
    "@machinat/core": "^0.5.0"
  }
}
