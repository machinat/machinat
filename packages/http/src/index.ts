export { default } from './module.js';
export { HttpConnector } from './Connector.js';
export * from './types.js';
