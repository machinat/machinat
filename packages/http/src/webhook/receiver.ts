import type { IncomingMessage, ServerResponse } from 'http';
import readRawBody from 'raw-body';
import type { RequestHandler, RoutingInfo } from '../types.js';
import type { WebhookHandler } from './types.js';

const RAW_BODY_OPTION = { encoding: true as const };

class WebhookReceiver {
  _handleWebhook: WebhookHandler;

  constructor(webhookHandler: WebhookHandler) {
    this._handleWebhook = webhookHandler;
  }

  handleRequestCallback(): RequestHandler {
    return this.handleRequest.bind(this);
  }

  async handleRequest(
    req: IncomingMessage,
    res: ServerResponse,
    routingInfo: RoutingInfo
  ): Promise<void> {
    try {
      let body: undefined | string;
      if (
        req.method === 'POST' ||
        req.method === 'PUT' ||
        req.method === 'DELETE' ||
        req.method === 'PATCH'
      ) {
        body = await readRawBody(req, RAW_BODY_OPTION);
      }

      const metadata = {
        source: 'webhook' as const,
        request: {
          method: req.method as string,
          url: req.url as string,
          headers: req.headers,
          body,
        },
      };

      const {
        code,
        headers,
        body: resBody,
      } = await this._handleWebhook(metadata, routingInfo);

      res.writeHead(code, headers);
      res.end(
        typeof resBody === 'string'
          ? resBody
          : resBody
          ? JSON.stringify(resBody)
          : undefined
      );
    } catch (err) {
      res.statusCode = 500;
      res.end(
        JSON.stringify({
          code: 500,
          message: err.message,
        })
      );
    }
  }
}

export default WebhookReceiver;
