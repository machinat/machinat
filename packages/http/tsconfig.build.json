{
  "extends": "../../tsconfig.build.json",
  "include": ["src/**/*.ts"],
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./lib",
    "baseUrl": ".",
    "paths": {
      "@sociably/core/*": ["../core/lib/*"]
    }
  },
  "references": [
    { "path": "../core/tsconfig.build.json" }
  ]
}
