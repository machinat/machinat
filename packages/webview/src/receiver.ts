import type { PopEventWrapper, PopEventFn, PopErrorFn } from '@machinat/core';
import { makeClassProvider } from '@machinat/core/service';
import ModuleUtilitiesI from '@machinat/core/base/ModuleUtilities';
import {
  AnyServerAuthenticator,
  UserOfAuthenticator,
  ContextOfAuthenticator,
} from '@machinat/auth';
import type { HttpRequestInfo } from '@machinat/http';
import {
  EventInput,
  EventValue,
  ConnectEventValue,
  DisconnectEventValue,
} from '@machinat/websocket';

import { WEBVIEW } from './constant';
import { SocketServerP, PlatformUtilitiesI } from './interface';
import { BotP } from './bot';
import { WebviewConnection } from './channel';
import { createEvent } from './utils';
import type { WebviewEventContext } from './types';

/**
 * @category Provider
 */
export class WebviewReceiver<
  Authenticator extends AnyServerAuthenticator,
  Value extends EventValue = EventValue
> {
  private _bot: BotP<Authenticator>;
  private _server: SocketServerP<Authenticator>;

  private _popEvent: PopEventFn<
    WebviewEventContext<Authenticator, Value>,
    null
  >;

  private _popError: PopErrorFn;

  constructor(
    bot: BotP<Authenticator>,
    server: SocketServerP<Authenticator>,
    popEventWrapper: PopEventWrapper<WebviewEventContext<Authenticator>, null>,
    popError: PopErrorFn
  ) {
    this._bot = bot;
    this._server = server;

    this._popEvent = popEventWrapper(() => Promise.resolve(null));
    this._popError = popError;

    this._server.on(
      'events',
      (values, { connId, user, request, authContext }) => {
        values.forEach((value) => {
          this._issueEvent(value, connId, user, request, authContext);
        });
      }
    );

    this._server.on('connect', ({ connId, user, request, authContext }) => {
      const value: ConnectEventValue = {
        category: 'connection',
        type: 'connect',
        payload: null,
      };
      this._issueEvent(value, connId, user, request, authContext);
    });

    this._server.on('disconnect', ({ reason }, connData) => {
      const { connId, user, request, authContext } = connData;
      const value: DisconnectEventValue = {
        category: 'connection',
        type: 'disconnect',
        payload: { reason },
      };

      this._issueEvent(value, connId, user, request, authContext);
    });

    this._server.on('error', (err: Error) => {
      this._popError(err);
    });
  }

  private _issueEvent(
    value: EventInput,
    connId: string,
    user: UserOfAuthenticator<Authenticator>,
    request: HttpRequestInfo,
    authContext: ContextOfAuthenticator<Authenticator>
  ) {
    const channel = new WebviewConnection(this._server.id, connId);
    this._popEvent({
      platform: WEBVIEW,
      bot: this._bot,
      event: createEvent(value, channel, user),
      metadata: {
        source: 'websocket',
        request,
        connection: channel,
        auth: authContext,
      },
      reply: (message) => this._bot.render(channel, message),
    }).catch(this._popError);
  }
}

export const ReceiverP = makeClassProvider({
  lifetime: 'singleton',
  deps: [BotP, SocketServerP, ModuleUtilitiesI, PlatformUtilitiesI],
  factory: (bot, server, { popError }, { popEventWrapper }) =>
    new WebviewReceiver(bot, server, popEventWrapper, popError),
})(WebviewReceiver);

export type ReceiverP<
  Authenticator extends AnyServerAuthenticator,
  Value extends EventValue = EventValue
> = WebviewReceiver<Authenticator, Value>;
