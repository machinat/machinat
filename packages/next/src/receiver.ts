import { parse as parseUrl } from 'url';
import { STATUS_CODES, IncomingMessage, ServerResponse } from 'http';
import { Socket as NetSocket } from 'net';
import {
  maybeInjectContainer,
  serviceProviderClass,
  createEmptyScope,
  ServiceScope,
} from '@sociably/core/service';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import type { PopErrorFn } from '@sociably/core';
import type {
  RequestHandler,
  UpgradeHandler,
  RoutingInfo,
} from '@sociably/http';
import { ServerI, ConfigsI } from './interface.js';
import type { NextServer, NextRequestHandler } from './types.js';

type NextReceiverOptions = {
  noPrepare?: boolean;
  handleRequest?: NextRequestHandler;
  initScope?: () => ServiceScope;
  popError?: PopErrorFn | null;
};

/**
 * @category Provider
 */
export class NextReceiver {
  private nextServer: NextServer;
  private defaultNextHandler: ReturnType<NextServer['getRequestHandler']>;

  private shouldPrepare: boolean;
  private isPrepared: boolean;

  private requestHandler: NextRequestHandler;
  private initScope: () => ServiceScope;
  private popError: (err: Error) => void;

  constructor(
    nextApp: NextServer,
    { noPrepare, handleRequest, initScope, popError }: NextReceiverOptions = {}
  ) {
    this.nextServer = nextApp;
    this.defaultNextHandler = nextApp.getRequestHandler();
    this.shouldPrepare = !noPrepare;

    this.requestHandler = handleRequest || (() => ({ ok: true }));
    this.initScope = initScope || createEmptyScope;
    this.popError =
      popError ||
      (() => {
        // do nothing
      });
  }

  async prepare(): Promise<void> {
    if (this.shouldPrepare) {
      await this.nextServer.prepare();
    }
    this.isPrepared = true;
  }

  async close(): Promise<void> {
    await this.nextServer.close();
  }

  handleRequest(
    req: IncomingMessage,
    res: ServerResponse,
    routing: RoutingInfo
  ): void {
    this.handleRequestImpl(req, res, routing).catch(this.popError);
  }

  handleRequestCallback(): RequestHandler {
    return this.handleRequest.bind(this);
  }

  handleHmrUpgrade(
    req: IncomingMessage,
    socket: NetSocket,
    _head: Buffer,
    { trailingPath }: RoutingInfo
  ): void {
    if (trailingPath !== '_next/webpack-hmr') {
      socket.write(
        `HTTP/1.1 404 Not Found\r\n` +
          'Connection: close\r\n' +
          'Content-Type: text/html\r\n' +
          `Content-Length: 9\r\n` +
          `\r\nNot Found`
      );
      socket.destroy();
    } else {
      this.defaultNextHandler(req, new ServerResponse(req)).catch(
        this.popError
      );
    }
  }

  handleHmrUpgradeCallback(): UpgradeHandler {
    return this.handleHmrUpgrade.bind(this);
  }

  private async handleRequestImpl(
    req: IncomingMessage,
    res: ServerResponse,
    { trailingPath }: RoutingInfo
  ) {
    if (!this.isPrepared) {
      res.writeHead(503);
      res.end(STATUS_CODES[503]);
      return;
    }

    const parsedUrl = parseUrl(req.url!, true);

    if (trailingPath.startsWith('_next')) {
      this.defaultNextHandler(req, res, parsedUrl);
      return;
    }

    try {
      const request = {
        method: req.method!,
        url: req.url!,
        route: trailingPath,
        headers: req.headers,
      };

      const response = await maybeInjectContainer(
        this.initScope(),
        this.requestHandler
      )(request);

      if (response.ok) {
        const { page, query, headers } = response;
        if (headers) {
          for (const [key, value] of Object.entries(headers)) {
            res.setHeader(key, value as string);
          }
        }

        if (page || query) {
          await this.nextServer.render(
            req,
            res,
            page || trailingPath,
            query || parsedUrl.query,
            parsedUrl
          );
        } else {
          this.defaultNextHandler(req, res, parsedUrl);
        }
      } else {
        const { code, reason, headers } = response;
        res.statusCode = code;

        if (headers) {
          for (const [name, value] of Object.entries(headers)) {
            if (value) {
              res.setHeader(name, value);
            }
          }
        }

        await this.nextServer.renderError(
          reason ? new Error(reason) : null,
          req,
          res,
          trailingPath,
          parsedUrl.query
        );
      }
    } catch (err) {
      this.popError(err);

      res.statusCode = 500;
      await this.nextServer.renderError(
        err,
        req,
        res,
        trailingPath,
        parsedUrl.query
      );
    }
  }
}

export const ReceiverP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [ServerI, ConfigsI, { require: ModuleUtilitiesI, optional: true }],
  factory: (nextApp, { noPrepare, handleRequest }, moduleUtils) =>
    new NextReceiver(nextApp, {
      noPrepare,
      handleRequest,
      initScope: moduleUtils?.initScope,
      popError: moduleUtils?.popError,
    }),
})(NextReceiver);

export type ReceiverP = NextReceiver;
