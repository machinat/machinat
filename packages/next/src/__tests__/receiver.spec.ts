import { parse as parseUrl } from 'url';
import {
  createEmptyScope,
  serviceContainer,
  ServiceScope,
} from '@sociably/core/service';
import moxy from '@moxyjs/moxy';
import type createNextApp from 'next';
import { IncomingMessage, ServerResponse } from 'http';
import { NextReceiver } from '../receiver';

const nextDefaultHandler = moxy(async () => {});

const nextApp = moxy<ReturnType<typeof createNextApp>>({
  getRequestHandler: () => nextDefaultHandler,
  prepare: async () => {},
  close: async () => {},
  render: async () => {},
  renderError: async () => {},
  setAssetPrefix() {},
  options: { assetPrefix: '' },
} as never);

const nextTick = () => new Promise((resolve) => process.nextTick(resolve));

const handleRequest = moxy(() => ({ ok: true as const }));
const initScope = moxy(createEmptyScope);
const popError = moxy();

const req = moxy<IncomingMessage>({
  method: 'GET',
  url: 'http://sociably.io/hello?foo=bar',
  headers: { 'x-y-z': 'abc' },
} as never);
let res;

beforeEach(() => {
  nextDefaultHandler.mock.clear();
  nextApp.mock.reset();
  req.mock.reset();
  res = moxy(new ServerResponse(req));

  handleRequest.mock.reset();
  initScope.mock.reset();
  popError.mock.reset();
});

it('respond 503 if request received before prepared', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: false,
    handleRequest,
  });

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(res.statusCode).toBe(503);
  expect(res.end).toHaveBeenCalledTimes(1);

  expect(nextApp.renderError).not.toHaveBeenCalled();
  expect(handleRequest).not.toHaveBeenCalled();
});

describe('.prepare()', () => {
  test('call nextServer.prepare()', async () => {
    const receiver = new NextReceiver(nextApp);
    expect(nextApp.prepare).not.toHaveBeenCalled();

    await receiver.prepare();

    expect(nextApp.prepare).toHaveBeenCalledTimes(1);
  });

  test('make no call if noPrepare is true', async () => {
    const receiver = new NextReceiver(nextApp, { noPrepare: true });
    await receiver.prepare();

    expect(nextApp.prepare).not.toHaveBeenCalled();
  });
});

test('.close() call nextServer.close()', async () => {
  const receiver = new NextReceiver(nextApp);
  await receiver.close();

  expect(nextApp.close).toHaveBeenCalledTimes(1);
});

test('handle http request', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(res.statusCode).toBe(200);

  expect(nextApp.renderError).not.toHaveBeenCalled();
  expect(nextApp.render).not.toHaveBeenCalled();

  expect(nextDefaultHandler).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler).toHaveBeenCalledWith(req, res, expect.any(Object));
  expect(nextDefaultHandler.mock.calls[0].args[2]).toMatchInlineSnapshot(`
    Url {
      "auth": null,
      "hash": null,
      "host": "sociably.io",
      "hostname": "sociably.io",
      "href": "http://sociably.io/hello?foo=bar",
      "path": "/hello?foo=bar",
      "pathname": "/hello",
      "port": null,
      "protocol": "http:",
      "query": Object {
        "foo": "bar",
      },
      "search": "?foo=bar",
      "slashes": true,
    }
  `);

  expect(handleRequest).toHaveBeenCalledTimes(1);
  expect(handleRequest).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://sociably.io/hello?foo=bar',
    route: '/hello',
    headers: { 'x-y-z': 'abc' },
  });
});

test('with entryPath', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://sociably.io/hello/world');
  receiver.handleRequest(req, res);
  await nextTick();

  expect(nextApp.renderError).not.toHaveBeenCalled();
  expect(nextApp.render).not.toHaveBeenCalled();

  expect(nextDefaultHandler).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler).toHaveBeenCalledWith(
    req,
    res,
    parseUrl('http://sociably.io/hello/world', true)
  );

  expect(handleRequest).toHaveBeenCalledTimes(1);
  expect(handleRequest).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://sociably.io/hello/world',
    route: '/world',
    headers: { 'x-y-z': 'abc' },
  });
});

it('respond 404 if entryPath not match', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://sociably.io/foo/world');

  receiver.handleRequest(req, res);
  await nextTick();

  expect(handleRequest).not.toHaveBeenCalled();
  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextDefaultHandler).not.toHaveBeenCalled();

  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextApp.renderError).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError).toHaveBeenCalledWith(
    null,
    req,
    res,
    '/foo/world',
    {}
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledTimes(1);
  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(404);
});

test('customized headers returned by handleRequest', async () => {
  handleRequest.mock.fake(async () => ({
    ok: true,
    headers: { 'x-foo': 'foo', 'x-bar': 'bar' },
  }));

  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(handleRequest).toHaveBeenCalledTimes(1);
  expect(handleRequest).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://sociably.io/hello?foo=bar',
    route: '/hello',
    headers: { 'x-y-z': 'abc' },
  });

  expect(res.statusCode).toBe(200);
  expect(res.getHeaders()).toEqual({ 'x-foo': 'foo', 'x-bar': 'bar' });
});

it('call next.render() with page/query returned by handleRequest', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  handleRequest.mock.fake(async () => ({
    ok: true,
    page: '/hello/world',
    query: { foo: 'bar' },
    headers: { 'x-y-z': 'a_b_c' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(handleRequest).toHaveReturnedTimes(1);
  expect(handleRequest.mock.calls[0].args[0]).toMatchSnapshot();

  expect(nextDefaultHandler).not.toHaveBeenCalled();
  expect(nextApp.renderError).not.toHaveBeenCalled();

  expect(nextApp.render).toHaveBeenCalledTimes(1);
  expect(nextApp.render).toHaveBeenCalledWith(
    req,
    res,
    '/hello/world',
    { foo: 'bar' },
    expect.any(Object)
  );

  expect(nextApp.render.mock.calls[0].args[4]).toMatchInlineSnapshot(`
    Url {
      "auth": null,
      "hash": null,
      "host": "sociably.io",
      "hostname": "sociably.io",
      "href": "http://sociably.io/hello?foo=bar",
      "path": "/hello?foo=bar",
      "pathname": "/hello",
      "port": null,
      "protocol": "http:",
      "query": Object {
        "foo": "bar",
      },
      "search": "?foo=bar",
      "slashes": true,
    }
  `);

  expect(handleRequest).toHaveBeenCalledTimes(1);
  expect(handleRequest).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://sociably.io/hello?foo=bar',
    route: '/hello',
    headers: { 'x-y-z': 'abc' },
  });

  expect(res.statusCode).toBe(200);
  expect(res.getHeaders()).toEqual({ 'x-y-z': 'a_b_c' });
});

test('use service container for handleRequest', async () => {
  const handleRequestFn = moxy(() => ({ ok: true as const, page: '/foo' }));
  const useRequestHandler = moxy(() => handleRequestFn);

  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    initScope,
    handleRequest: serviceContainer({ deps: [ServiceScope] })(
      useRequestHandler
    ),
  });
  await receiver.prepare();

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(initScope).toHaveBeenCalledTimes(1);
  expect(useRequestHandler).toHaveReturnedTimes(1);
  expect(useRequestHandler).toHaveBeenCalledWith(
    initScope.mock.calls[0].result
  );

  expect(nextDefaultHandler).not.toHaveBeenCalled();
  expect(nextApp.renderError).not.toHaveBeenCalled();

  expect(nextApp.render).toHaveBeenCalledTimes(1);
  expect(nextApp.render).toHaveBeenCalledWith(
    req,
    res,
    '/foo',
    { foo: 'bar' },
    expect.any(Object)
  );

  expect(handleRequestFn).toHaveBeenCalledTimes(1);
  expect(handleRequestFn).toHaveBeenCalledWith({
    method: 'GET',
    url: 'http://sociably.io/hello?foo=bar',
    route: '/hello',
    headers: { 'x-y-z': 'abc' },
  });

  expect(res.statusCode).toBe(200);
});

it('call next.renderError() handleRequest return not ok', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  handleRequest.mock.fake(async () => ({
    ok: false,
    code: 418,
    reason: "I'm a teapot",
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(handleRequest).toHaveBeenCalledTimes(1);
  expect(handleRequest.mock.calls[0].args[0]).toMatchSnapshot();

  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextDefaultHandler).not.toHaveBeenCalled();

  expect(nextApp.renderError).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(418);
  expect(popError).not.toHaveBeenCalled();
});

it('call next.renderError() with customized headers return by handleRequest', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  handleRequest.mock.fake(async () => ({
    ok: false,
    code: 418,
    reason: "I'm a teapot",
    headers: { 'x-x-x': 't-e-a-p-o-t' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(nextApp.renderError).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledTimes(1);
  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(418);
  expect(res.setHeader).toHaveBeenCalledTimes(1);
  expect(res.setHeader).toHaveBeenCalledWith('x-x-x', 't-e-a-p-o-t');
});

it('pass "_next" api calls to next directly', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest,
  });
  await receiver.prepare();

  req.mock.getter('url').fakeReturnValue('http://sociably.io/hello/_next/xxx');

  receiver.handleRequest(req, res);
  await nextTick();

  expect(handleRequest).not.toHaveBeenCalled();
  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextApp.renderError).not.toHaveBeenCalled();

  expect(nextDefaultHandler).toHaveBeenCalledTimes(1);
  expect(nextDefaultHandler).toHaveBeenCalledWith(req, res, expect.any(Object));
  const parsedUrl = nextDefaultHandler.mock.calls[0].args[2];
  expect(parsedUrl.path).toBe('/hello/_next/xxx');
  expect(parsedUrl.pathname).toBe('/hello/_next/xxx');
});

test('entryPath does not affect page params from middlewares', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest,
  });
  await receiver.prepare();

  handleRequest.mock.fake(async () => ({
    ok: true,
    page: '/hello/world',
    query: { foo: 'bar' },
  }));

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(nextApp.render).toHaveBeenCalledTimes(1);
  expect(nextApp.render).toHaveBeenCalledWith(
    req,
    res,
    '/hello/world',
    { foo: 'bar' },
    expect.any(Object)
  );
});

it('call next.renderError() if middlewares reject', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    popError,
    handleRequest,
  });
  await receiver.prepare();

  handleRequest.mock.fake(async () => {
    throw new Error("I'm a teapot");
  });

  expect(receiver.handleRequest(req, res)).toBe(undefined);
  await nextTick();

  expect(handleRequest).toHaveBeenCalledTimes(1);

  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextDefaultHandler).not.toHaveBeenCalled();

  expect(nextApp.renderError).toHaveBeenCalledTimes(1);
  expect(nextApp.renderError).toHaveBeenCalledWith(
    new Error("I'm a teapot"),
    req,
    res,
    '/hello',
    { foo: 'bar' }
  );

  expect(res.mock.setter('statusCode')).toHaveBeenCalledWith(500);

  expect(popError).toHaveBeenCalledTimes(1);
  expect(popError).toHaveBeenCalledWith(new Error("I'm a teapot"));
});

it('pass hmr upgrade request to next server', async () => {
  const receiver = new NextReceiver(nextApp, {
    noPrepare: true,
    entryPath: '/hello',
    popError,
    handleRequest,
  });
  await receiver.prepare();

  const socket = moxy({ write: () => {}, destroy: () => {} });

  receiver.handleHmrUpgrade(req, socket as never, Buffer.from(''), {
    originalPath: '/hello/_next/webpack-hmr',
    matchedPath: '/hello',
    trailingPath: '_next/webpack-hmr',
  });
  await nextTick();

  expect(nextDefaultHandler).toHaveBeenCalledWith(
    req,
    expect.any(ServerResponse)
  );
  expect(socket.write).not.toHaveBeenCalled();
  expect(socket.destroy).not.toHaveBeenCalled();

  receiver.handleHmrUpgrade(req, socket as never, Buffer.from(''), {
    originalPath: '/hello/_next/wrong-path',
    matchedPath: '/hello',
    trailingPath: '_next/wrong-path',
  });
  await nextTick();

  expect(socket.destroy).toHaveBeenCalledTimes(1);
  expect(socket.write).toHaveBeenCalledTimes(1);
  expect(socket.write.mock.calls[0].args[0]).toMatchInlineSnapshot(`
    "HTTP/1.1 404 Not Found
    Connection: close
    Content-Type: text/html
    Content-Length: 9

    Not Found"
  `);

  expect(nextDefaultHandler).toHaveBeenCalledTimes(1);
  expect(handleRequest).not.toHaveBeenCalled();
  expect(nextApp.render).not.toHaveBeenCalled();
  expect(nextApp.renderError).not.toHaveBeenCalled();
});
