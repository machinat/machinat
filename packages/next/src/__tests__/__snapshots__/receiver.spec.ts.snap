// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call next.render() with page/query returned by handleRequest 1`] = `
Object {
  "headers": Object {
    "x-y-z": "abc",
  },
  "method": "GET",
  "route": "/hello",
  "url": "http://sociably.io/hello?foo=bar",
}
`;

exports[`call next.renderError() handleRequest return not ok 1`] = `
Object {
  "headers": Object {
    "x-y-z": "abc",
  },
  "method": "GET",
  "route": "/hello",
  "url": "http://sociably.io/hello?foo=bar",
}
`;
